include/geode/basic/database.h

```diff
@@ -0,0 +1,164 @@
+/*
+ * Copyright (c) 2019 - 2022 Geode-solutions
+ *
+ * Permission is hereby granted, free of charge, to any person obtaining a copy
+ * of this software and associated documentation files (the "Software"), to deal
+ * in the Software without restriction, including without limitation the rights
+ * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+ * copies of the Software, and to permit persons to whom the Software is
+ * furnished to do so, subject to the following conditions:
+ *
+ * The above copyright notice and this permission notice shall be included in
+ * all copies or substantial portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+ * SOFTWARE.
+ *
+ */
+
+#pragma once
+
+#include <geode/basic/common.h>
+#include <geode/basic/logger.h>
+#include <geode/basic/pimpl.h>
+
+#include <geode/basic/identifier.h>
+#include <geode/basic/uuid.h>
+
+namespace geode
+{
+    class Identifier;
+    struct uuid;
+} // namespace geode
+
+namespace geode
+{
+    /*!
+     * Stores any classes inherited from Identifier. It owns every data
+     * registered. Data is also stored on disk and offload from the memory after
+     * some unused time to save memory and performances.
+     */
+    class opengeode_basic_api Database
+    {
+        OPENGEODE_DISABLE_COPY( Database );
+        struct Storage;
+
+    public:
+        using serializer_function = std::function< void( PContext& ) >;
+
+        /*!
+         * Classe holding a const reference of a data.
+         * @warning Do not destroy this Data class before the const reference
+         * obtained using its get() method is no longer in used
```

Comment:
```suggestion
         * @warning Do not destroy this Data class if the const reference
         * obtained using its get() method is still used
```
else? what's happened if it is done?

Reply:
Do we really need to document what happen if you di not respect the rule?

Reply:
No but can you tell me, just to know for future debug?

Reply:
when all the Data objects are deleted, the database starts a countdown to offload the actual data from the memory. So if you destroy your Data object but are still using the referenced data, you might get segfault or weird bugs because the memory has been cleaned.